flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/PubSubSource.java

```diff
@@ -0,0 +1,296 @@
+import com.google.pubsub.v1.ProjectSubscriptionName;
+import com.google.pubsub.v1.PubsubMessage;
+import com.google.pubsub.v1.PullRequest;
+import com.google.pubsub.v1.PullResponse;
+import com.google.pubsub.v1.ReceivedMessage;
+import io.grpc.netty.shaded.io.netty.channel.EventLoopGroup;
+import io.grpc.netty.shaded.io.netty.channel.nio.NioEventLoopGroup;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.cloud.pubsub.v1.SubscriptionAdminSettings.defaultCredentialsProviderBuilder;
+
+/**
+ * PubSub Source, this Source will consume PubSub messages from a subscription and Acknowledge them on the next checkpoint.
+ * This ensures every message will get acknowledged at least once.
+ */
+public class PubSubSource<OUT> extends MultipleIdsMessageAcknowledgingSourceBase<OUT, String, String>
+		implements ResultTypeQueryable<OUT>, ParallelSourceFunction<OUT>, StoppableFunction {
+	private static final Logger LOG = LoggerFactory.getLogger(PubSubSource.class);
+	protected final DeserializationSchema<OUT> deserializationSchema;
+	protected final PubSubSubscriberFactory pubSubSubscriberFactory;
+	protected final Credentials credentials;
+	protected final String projectSubscriptionName;
+	protected final int maxMessagesPerPull;
+
+	protected transient boolean deduplicateMessages;
+	protected transient SubscriberStub subscriber;
+	protected transient PullRequest pullRequest;
+	protected transient EventLoopGroup eventLoopGroup;
+
+	protected transient volatile boolean isRunning;
+	protected transient volatile ApiFuture<PullResponse> messagesFuture;
+
+	PubSubSource(DeserializationSchema<OUT> deserializationSchema, PubSubSubscriberFactory pubSubSubscriberFactory, Credentials credentials, String projectSubscriptionName, int maxMessagesPerPull) {
+		super(String.class);
+		this.deserializationSchema = deserializationSchema;
+		this.pubSubSubscriberFactory = pubSubSubscriberFactory;
+		this.credentials = credentials;
+		this.projectSubscriptionName = projectSubscriptionName;
+		this.maxMessagesPerPull = maxMessagesPerPull;
+	}
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		super.open(configuration);
+		if (hasNoCheckpointingEnabled(getRuntimeContext())) {
+			throw new IllegalArgumentException("The PubSubSource REQUIRES Checkpointing to be enabled and " +
+				"the checkpointing frequency must be MUCH lower than the PubSub timeout for it to retry a message.");
+		}
+
+		getRuntimeContext().getMetricGroup().gauge("PubSubMessagesProcessedNotAcked", this::getOutstandingMessagesToAck);
+
+		this.eventLoopGroup = new NioEventLoopGroup();
+		this.subscriber = pubSubSubscriberFactory.getSubscriber(eventLoopGroup, credentials);
+		this.deduplicateMessages = getRuntimeContext().getNumberOfParallelSubtasks() == 1;
```

Comment:
See @rmetzger comments about this a week or so ago. I'll resolve this review comment because I am losing track, please do add your thought to that `conversation` :)